src/shared/image-policy.c

```diff
@@ -0,0 +1,653 @@
+
+#include "alloc-util.h"
+#include "extract-word.h"
+#include "image-policy.h"
+#include "sort-util.h"
+#include "string-util.h"
+#include "strv.h"
+
+/* Rationale for the chosen syntax:
+ *
+ * → one line, so that it can be reasonably added to a shell command line, for example via `systemd-dissect
+ *   --image-policy=…` or to the kernel command line via `systemd.image_policy=`.
+ *
+ * → no use of "," or ";" as separators, so that it can be included in mount/fstab-style option strings and
+ *   doesn't require escaping. Instead, separators are ":", "=", "+" which should be fine both in shell
+ *   command lines and in mount/fstab style option strings.
+ */
+
+static int partition_policy_compare(const PartitionPolicy *a, const PartitionPolicy *b) {
+        return CMP(ASSERT_PTR(a)->designator, ASSERT_PTR(b)->designator);
+}
+
+static PartitionPolicy* image_policy_bsearch(const ImagePolicy *policy, PartitionDesignator designator) {
+        if (!policy)
+                return NULL;
+
+        return typesafe_bsearch(
+                        &(PartitionPolicy) { .designator = designator },
+                        ASSERT_PTR(policy)->policies,
+                        ASSERT_PTR(policy)->n_policies,
+                        partition_policy_compare);
+}
+
+static PartitionPolicyFlags partition_policy_normalized_flags(const PartitionPolicy *policy) {
+        PartitionPolicyFlags flags = ASSERT_PTR(policy)->flags;
+
+        /* If no protection flag is set, then this means all are set */
+        if ((flags & _PARTITION_POLICY_USE_MASK) == 0)
+                flags |= PARTITION_POLICY_OPEN;
+
+        /* If this is a verity or verity signature designator, then mask off all protection bits, this after
+         * all needs no protection, because it *is* the protection */
+        if (partition_verity_to_data(policy->designator) >= 0 ||
+            partition_verity_sig_to_data(policy->designator) >= 0)
+                flags &= ~(PARTITION_POLICY_VERITY|PARTITION_POLICY_SIGNED|PARTITION_POLICY_ENCRYPTED);
+
+        /* if this designator has no verity concept, then mask off verity protection flags */
+        if (partition_verity_of(policy->designator) < 0)
+                flags &= ~(PARTITION_POLICY_VERITY|PARTITION_POLICY_SIGNED);
+
+        if ((flags & _PARTITION_POLICY_USE_MASK) == PARTITION_POLICY_ABSENT)
+                /* If the partition must be absent, then the gpt flags don't matter */
+                flags &= ~(_PARTITION_POLICY_READ_ONLY_MASK|_PARTITION_POLICY_GROWFS_MASK);
+        else {
+                /* If the gpt flags bits are not specified, set both options for each */
+                if ((flags & _PARTITION_POLICY_READ_ONLY_MASK) == 0)
+                        flags |= PARTITION_POLICY_READ_ONLY_ON|PARTITION_POLICY_READ_ONLY_OFF;
+                if ((flags & _PARTITION_POLICY_GROWFS_MASK) == 0)
+                        flags |= PARTITION_POLICY_GROWFS_ON|PARTITION_POLICY_GROWFS_OFF;
```

Comment:
well, the thing is we want to be able to declare one of three policies:

1. growfs flag must be on
2. growfs flag must be off
3. dont care

to encode 3 options we need at least 2 bits, which actually makes 2^2 = 4 options. So how do we map 3 options to 4 values? i simply figured it would make most sense to encode it like this, and define them so that:

1 0 → growfs must be on
0 1 → growfs must be off
1 1 → growfs must be either on, or off (aka "don't care"), explicitly configured by user
0 0 → growfs must be either on, or off (aka "don't care"), not configured explicitly by user, but implied because left unspecified

The function above then removes the destinction between the latter two cases: it just converts the "user doesn't care and left unspecified" and "user explicitly declared doesn't matter" into a  generic "doesn't matter".

this also has the nice benefit that on one hand you can quickly check if it's ok that the read-only gpt flag is on by doing a `flags & PARTITION_POLICY_READ_ONLY_ON` check, and if you want to check if it's ok that it is off, you can do `flags & PARTITION_POLICY_READ_ONLY_OFF` and you get the right results.

or to say this differently: when parsing/processing policies it's quite interesting to distinguish what the user explicitly set to "don't care" and what they just left out of the policy and thus gets an implicit "don't care". But once you enforce the policy you don#t care about the distinction anymore, and that's what this function here does: it normalizes the differences away, so that we get the same effective policy in both cases, and thus can check more easily by just checking presence of simple flags.